### PR TITLE
Run integration tests on circleci

### PR DESCRIPTION
**Pull request by rodrigosaito**
This PR is to address the following issue: https://github.com/go-kit/kit/issues/328

This is still working in progress, I'd be happy to finish this if it's looking good.

There are some workarounds to fix the GOPATH in circle.yml, it might not be necessary on the original repo.

### CODE REVIEW

File: sd/consul/integration_test.go [hunk 1]
```diff
@@ -1,4 +1,4 @@
-// +build integration
+// +build integration-ignore
```
Comment (peterbourgon) [inline sd/consul/integration_test.go]: What does this do?

File: circle.yml [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+  pre:
+    - curl -sSL https://s3.amazonaws.com/circle-downloads/install-circleci-docker.sh | bash -s -- 1.10.0
+  services:
+    - docker
+
+dependencies:
+  pre:
+    - sudo pip install docker-compose
+    - docker-compose -f docker-compose-integration.yml up -d --force-recreate
+    - mkdir -p ${HOME}/.go_workspace/src/github.com/go-kit/ && ln -sf ${HOME}/kit ${HOME}/.go_workspace/src/github.com/go-kit/
```
Comment (peterbourgon) [inline circle.yml]: Why is this necessary?

File: circle.yml [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+  override:
+    - cd ${HOME}/.go_workspace/src/github.com/go-kit/kit && go get -t -d -v ./...
```
Comment (peterbourgon) [inline circle.yml]: Shouldn't it be sufficient to `go get -d -t -v ./...` without the cd?

File: sd/consul/integration_test.go (hunk 1 above)
Comment (rodrigosaito) [inline sd/consul/integration_test.go]: just ignoring those tests for now, will remove that once everything is done.

File: circle.yml (hunk 3 above)
Comment (rodrigosaito) [inline circle.yml]: yes, probably with go-kit/kit repo, not with my fork.

before sending the final PR I will remove all this workaround that I need to run builds on circle from my fork.

**Comment by peterbourgon** [comment]
Ping? :)

**Comment by rodrigosaito** [comment]
@peterbourgon sorry about the delay.

so should I keep going with this one?

**Comment by peterbourgon** [comment]
No worries at all! Yes, please, if you can; otherwise I'll take it over :)

**Comment by rodrigosaito** [comment]
Yes, I can and will be happy to help. Will try to finish this by the weekend.

**Comment by peterbourgon** [comment]
Friendly ping :) Any update here?

**Comment by rodrigosaito** [comment]
I am being busy lately, will try to find some time to work on this in the next few days.

**Comment by rodrigosaito** [comment]
@peterbourgon done!

the tests are failing now, probably because circleci is configured to use the old image (12.04) instead of the new one (14.04). If you change that, then the build should pass. Another solution is to stop zookeeper on circle.yml, I can do that if you prefer this solution.

**Comment by peterbourgon** [comment]
Well, the error is pretty explicit; the docker-compose up fails with

```
ERROR: for zk  Cannot start service zk: failed to create endpoint kit_zk_1 on network kit_default: Error starting userland proxy: listen tcp 0.0.0.0:2181: bind: address already in use
```

When I try to do the same thing locally, I get a similar error:

```
ERROR: for etcd  Cannot start service etcd: driver failed programming external connectivity on endpoint kit_etcd_1 (0577dd66f5552c29e6bf8fe0c882c9cfcf4515215a1d2bdf19650df0b44dd90e): Error starting userland proxy: listen tcp 0.0.0.0:2379: bind: address already in use
```

I think there must be some error in the way you've got the docker-compose file orchestrating things. Does this work for you locally?

**Comment by rodrigosaito** [comment]
The error happens on circle ci, probably because it's running on 12.04 image which has the zookeeper up and running(https://circleci.com/docs/build-image-precise/#databases), conflicting the ports as the docker-compose is using default ports for every key store (consul, etcd and zk). I tested on circle ci using 14.04 (https://circleci.com/docs/build-image-trusty/#databases) which doesn't have a zookeeper running.

On your machine you might have a etcd running which will also conflict the ports.

Yes it works locally for me and when I run on circle ci from my repo.

I can change to use random ports if you prefer.

**Comment by peterbourgon** [comment]
Ah, OK! Thanks for the more lengthy explanation. You're probably right; I was running that in a minikube (Kubernetes) VM which probably has etcd listening on the default port. I'll poke around a bit more.

**Comment by rodrigosaito** [comment]
hey @peterbourgon any update on this? do you need some help?

**Comment by peterbourgon** [comment]
Thanks for the poke :) I got it working locally. A few changes I'd like to see to the PR itself incoming...

File: circle.yml [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+    - sudo pip install docker-compose
+    - docker-compose -f docker-compose-integration.yml up -d --force-recreate
+
+test:
+  override:
+    - go test -v -race -tags integration ./...:
+        environment:
+          ETCD_ADDR: http://localhost:2379
+          CONSUL_ADDR: localhost:8500
+          ZK_ADDR: localhost:2181
```
Comment (peterbourgon) [inline circle.yml]: Can you make whatever changes are necessary to get this working on Circle in the file, or do I need to change something in the repo settings?

File: sd/zk/integration_test.go [hunk 5]
```diff
@@ -17,22 +16,12 @@ var (
...
-func TestMain(m *testing.M) {
-	flag.Parse()
-
-	fmt.Println("Starting ZooKeeper server...")
-
-	ts, err := stdzk.StartTestCluster(1, nil, nil)
-	if err != nil {
-		fmt.Printf("ZooKeeper server error: %v\n", err)
-		os.Exit(1)
+func init() {
```
Comment (peterbourgon) [inline sd/zk/integration_test.go]: Why was this moved out of TestMain and into init?

Comment (rodrigosaito) [inline sd/zk/integration_test.go]: removed the code that was starting zookeeper (this previously needed the zk jar file somewhere). Not really sure the difference of init and TestMain. Would you prefer to have this code on TestMain instead?

Comment (rodrigosaito) [inline sd/zk/integration_test.go]: ok, got it. Will do that later today.

Comment (rodrigosaito) [inline sd/zk/integration_test.go]: done!

File: circle.yml (hunk 4 above)
Comment (peterbourgon) [inline circle.yml]: Done. Takes a fresh commit to test; we'll do it on master! ;)

**Comment by peterbourgon** [comment]
Brilliant! Works locally and I'm happy with the delta. Thanks very much for the contribution! 👍
